src/composeWithMongooseDiscriminators.js

```diff
@@ -0,0 +1,379 @@
+    const childFieldName = childFields.find(fld => fld === field);
+
+    if (childFieldName) {
+      childTC.extendField(field, {
+        type: baseDTC.getFieldType(field),
+      });
+    } else {
+      childTC.setField(field, baseDTC.getField(field));
+    }
+  }
+
+  return childTC;
+}
+
+function getBaseTCFieldsWithTypes(baseTC: TypeComposer) {
+  const baseFields = baseTC.getFieldNames();
+  const baseFieldsWithTypes: GraphQLFieldConfigMap<any, any> = {};
+
+  for (const field of baseFields) {
+    baseFieldsWithTypes[field] = {
+      type: baseTC.getFieldType(field),
+    };
+  }
+
+  return baseFieldsWithTypes;
+}
+
+function createDInterface(baseModelTC: DiscriminatorTypeComposer): GraphQLInterfaceType {
+  return new GraphQLInterfaceType({
+    name: baseModelTC.getDBaseName(),
+
+    resolveType: (value: any) => {
+      const childDName = value[baseModelTC.getDKey()];
+      const sComposer = baseModelTC.getGQC(); // get schemaComposer
+
+      if (childDName) {
+        return sComposer.getTC(childDName).getType();
+      }
+
+      // as fallback return BaseModelTC
+      return sComposer.getTC(baseModelTC.getTypeName()).getType();
+    },
+    // hoisting issue solved, get at time :)
+    fields: () => getBaseTCFieldsWithTypes(baseModelTC),
+  });
+}
+
+export class DiscriminatorTypeComposer extends TypeComposer {
+  modelName: string;
+
+  discriminatorKey: string;
+
+  DKeyETC: EnumTypeComposer;
+
+  GQC: SchemaComposer<any>;
+
+  opts: Options;
+
+  DInterface: GraphQLInterfaceType;
+
+  discriminators: Discriminators;
+
+  CDTCs: Array<ChildDiscriminatorTypeComposer>;
+
+  constructor(
+    baseModel: Model,
+    opts: Options = {
+      reorderFields: true,
+      customizationOptions: {},
+    }
+  ) {
+    if (!baseModel || !(baseModel: any).discriminators) {
+      throw Error('Discriminator Key not Set, Use composeWithMongoose for Normal Collections');
+    }
+
+    super(composeWithMongoose(baseModel, opts.customizationOptions).gqType);
+
+    // !ORDER MATTERS
+    this.opts = opts || {};
+    this.modelName = (baseModel: any).modelName;
+    this.discriminatorKey = (baseModel: any).schema.get('discriminatorKey') || '__t';
+
+    // discriminants and object containing all discriminants with key
+    // key being their DNames
+    this.discriminators = (baseModel: any).discriminators;
+
+    this.CDTCs = [];
+    this.GQC = opts.customizationOptions.schemaComposer || schemaComposer;
+    this.setTypeName(`Generic${this.modelName}`);
+    this.DKeyETC = createAndSetDKeyETC(this, this.discriminators);
+
+    reorderFields(this, this.opts.reorderFields);
+
+    this.DInterface = createDInterface(this);
+    this.setInterfaces([this.DInterface]);
+
+    // Add a Generic Field, else we have generic Errors when resolving interface
+    // this is somehow i don't understand, but we don't get any type if we never query it
+    // I guess under the hud, graphql-compose shakes it off.
+    this.GQC.Query.addFields({
```

Comment:
It's very bad to automatically add fields to Query type. Developers should to it manually in their code. In my schema I'm using resolvers under `my`, `admin`, `public` fields in the Query (they have different permissions and a different set of resolvers (eg. I do not use pagination, remove* resolvers under `Query.public`)

Reply:
Will work on removing.

Reply:
@nodkz, i have this issue with the discriminators. When you don't add a particular discriminator type to query, mutation or subscription, when resolving the Interface, Graphql throws an Error if it references that type which is not in schema.

```javascript
buildSchema(): GraphQLSchema {
    const roots = {};

    if (this.has('Query')) {
      const tc = this.getTC('Query');
      this.removeEmptyTypes(tc, new Set());
      roots.query = tc.getType();
    }

    if (this.has('Mutation')) {
      const tc = this.getTC('Mutation');
      this.removeEmptyTypes(tc, new Set());
      roots.mutation = tc.getType();
    }

    if (this.has('Subscription')) {
      const tc = this.getTC('Subscription');
      this.removeEmptyTypes(tc, new Set());
      roots.subscription = tc.getType();
    }

    if (Object.keys(roots).length === 0) {
      throw new Error(
        'Can not build schema. Must be initialized at least one ' +
          'of the following types: Query, Mutation, Subscription.'
      );
    }

    return new GraphQLSchema(roots);
  }
```
While seeking you opinion on how to solve the issue, i am proposing to add a field called `_disTypes` that contains all types that are either Discrimator or Child of Discriminator to `rootQuery`. It only host the Types, cannot be Queried, nor Mutated.

### In composeWithMongooseDiscriminators.js

```javascript
if (!this.GQC.rootQuery().hasField('_disTypes')) {
      this.GQC.rootQuery().addFields({
        _disTypes: TypeComposer.create({
          name: '_disTypes',
          description: 'Hoisting for Discriminator Types',
        }),
    });
}

this.GQC.rootQuery()
    .getFieldTC('_disTypes')
    .addFields({
      [this.getTypeName()]: this,
    });
```

Reply:
Can you add a broken test case without _distTypes logic? I can solve this problem today after several hours. I'm very dislike any pollution to the users' schemas.

Reply:
Added the test case, there are three test cases, 

### in src/__tests__/_disTypes-test.js

1. Test that doesn't include _distTypes but manually i manually added resolvers that return all the Types that implement DInterface
1. Test that includes _disTypes by setting discriminator `test_disTypes` option to `true`. This case executes without errors
1. Test that turns `test_disTypes` option to `false`, the case returns GraphqlError that i `asserted`.

Reply:
Your tests help me very well to figure out the problem.

I found a solution. But it required to wait for some time while I make proper changes in `graphql-compose` package. 

Just added [InterfaceTypeComposer](https://github.com/graphql-compose/graphql-compose/compare/interface-type-composer). It still in progress in a separate branch.

After that, it is necessary to teach `TypeComposer` to work with `InterfaceTypeComposer`.
And then make small changes in SchemaComposer. Midnight is already, try to do it tomorrow.

---

@mernxl while you wait for me, you may suggest some new methods to `graphql-compose` which it would be good to have, based on the experience of writing discriminators. If there are any ;)

Also, you may add some info about how discriminators work to README.md.

Reply:
@nodkz, great work on `InterfaceTypeComposer`. Don't know of those new methods as of now, but if they do pop up i will do a PR. 

* Will be adding the info on how discriminators work to README.md.

* While working on discriminators, i wrote a small function that merges two `TypeConverterOptions`, that is takes the first option and merges it with the second. The first opts only replaces the undefined opts in the second arg. I think it will be nice to expose this function so someone could create base type converter options and only merge them with extras when he needs them. 

```javascript
const baseCustomOptions: TypeConverterOpts = {
    fields: baseFields,
    inputType: {
      name: 'BaseInput',
      description: 'Hello Base',
      fields: baseInputTypeFields,
    },
    resolvers: {
      findMany: {
        limit: { defaultValue: 20 },
        // sort: false,
        skip: false,
        filter: {
          isRequired: true,
          removeFields: ['id', 'dob'],
          operators: {
            one: ['gt', 'gte', 'lt'],
            two: ['gt', 'gte', 'lt', 'in[]', 'nin[]'],
          },
        },
      },
      findById: false,
    },
  };

  const childCustomOptions: TypeConverterOpts = {
    fields: childFields,
    inputType: {
      name: 'ChildInputs',
      description: 'Hello Child',
      fields: childInputTypeFields,
    },
    resolvers: {
      findMany: {
        limit: { defaultValue: 50 },
        sort: false,
        // skip: false,
        filter: {
          removeFields: ['gender', 'dob', 'age'],
          operators: {
            one: ['gt', 'lte', 'ne', 'in[]', 'nin[]'],
            two: ['gt', 'gte', 'lt', 'lte', 'ne'],
            three: ['gte', 'lt'],
          },
        },
      },
      updateById: {
        input: {
          removeFields: ['one', 'two', 'five'],
          requiredFields: ['eight', 'two', 'five'],
        },
      },
    },
  };

  const expected: TypeConverterOpts = {
    fields: expectedFields,
    inputType: {
      name: 'ChildInputs',
      description: 'Hello Child',
      fields: expectedInputTypes,
    },
    resolvers: {
      findMany: {
        limit: { defaultValue: 50 },
        sort: false,
        skip: false,
        filter: {
          isRequired: true,
          removeFields: ['id', 'dob', 'gender', 'age'],
          operators: {
            one: ['gt', 'gte', 'lt', 'lte', 'ne', 'in[]', 'nin[]'],
            two: ['gt', 'gte', 'lt', 'in[]', 'nin[]', 'lte', 'ne'],
            three: ['gte', 'lt'],
          },
        },
      },
      findById: false,
      updateById: {
        input: {
          removeFields: ['one', 'two', 'five'],
          requiredFields: ['eight', 'two', 'five'],
        },
      },
    },
  };

  it('should merge customisation Options', () => {
    expect(mergeCustomizationOptions(baseCustomOptions, childCustomOptions)).toEqual(expected);
  });
```

Reply:
Sure, lets expose `mergeCustomizationOptions` method in this package.